### PR TITLE
Implement request throttling for bulk enrollment API (iss. #190)

### LINKED ISSUE #190: avoid/recover from Canvas API rate-limit errors in bulk enrollment API
> Using the Got retry feature, this implementation is often (usually?), but not always, successful in enrolling a full set of 400 students. It's ignoring the Canvas API throttling and essentially using brute force. From team discussions, this is the technique that I was directed to pursue. It may be important to revisit this to use request throttling before merging this PR. I don't know, that's up to the reviewers and testers to determine. I've received conflicting directions on this matter.

The quoted text above is from the description of this PR.  That is what I referred to in the meeting this morning when I brought up the subject of possibly needing request throttling for this feature.

_Originally posted by @lsloan in https://github.com/tl-its-umich-edu/canvas-course-manager-next/issues/165#issuecomment-921114376_

### PR DESCRIPTION
**Pull request by lsloan**
Resolves #190.

Attempting:

1. Use Got `calculatedDelay` setting of `retry` feature
   1. Use a static value
   2. Calculate a value based on retry count, Canvas response headers, etc.
2. Use proactive delays in the request resolution loop

### CODE REVIEW

**Comment by lsloan** [comment]
@ssciolla and @pushyamig, I've requested both of your reviews on this.  Given that Sam and I had different results using nearly the same code and we don't know why, this needs to be tested by multiple people.  In fact, if you think of someone else who would be able to test this in addition to you, that would be helpful.  Maybe Matt, Zhen, or Chris?

File: ccm_web/server/src/canvas/canvas.service.ts [hunk 1]
```diff
@@ -27,34 +27,35 @@ interface IncomingRateLimitedCanvasHttpHeaders extends IncomingHttpHeaders {
...
 const requestorOptions: GotOptions = {
   retry: {
-    limit: 2,
+    limit: 3,
     methods: ['POST', 'GET', 'PUT', 'DELETE'],
-    statusCodes: got.defaults.options.retry.statusCodes.concat([403])
+    statusCodes: got.defaults.options.retry.statusCodes.concat([403]),
+    calculateDelay: ({ attemptCount, retryOptions, error, computedValue }) => {
+      const headers = error.response?.headers as IncomingRateLimitedCanvasHttpHeaders
+      const delay: number = computedValue === 0 ? 0 : 5000
```
Comment (ssciolla) [inline ccm_web/server/src/canvas/canvas.service.ts:35]: I did some more testing with your code, and I encountered what I think you are seeing a few times. Basically you get receive `error` that isn't an `HTTPError` and doesn't have a status code but doesn't have a `code` either (like `ECONNRESET`). I'm not sure our configuration is set up to retry when this occurs.

There's another issue I saw that in these cases, you're still trying to access these custom headers, so there's an error message to the effect of the "property x-rate-limit-remaining doesn't exist on object undefined". This may be why you're not seeing `calculateDelay` in these cases, but I'm not sure about that.

Anyways, I'm going to do some more testing on Monday with something like this:
```
  retry: {
    limit: 3,
    methods: ['POST', 'GET', 'PUT', 'DELETE'],
    statusCodes: got.defaults.options.retry.statusCodes.concat([403]),
    calculateDelay: ({ attemptCount, retryOptions, error, computedValue }) => {
      const headers = error.response?.headers
      const isStrangeError = !(error instanceof HTTPError) && error.code === undefined
      logger.debug(`isStrangeError: ${String(isStrangeError)}`)
      const delay = computedValue === 0 && !isStrangeError ? 0 : 5000
      logger.debug(
        `calculateDelay [${String(attemptCount)}] — ` +
        `"delay": "${String(delay)}"; ` +
        `"retryOptions": "${JSON.stringify(retryOptions)}";` +
        `error.code: ${String(error.code)}`
      )
      if (headers !== undefined) {
        logger.debug(
          `"x-rate-limit-remaining": "${String(headers['x-rate-limit-remaining'])}"; ` +
          `"x-request-cost": "${String(headers['x-request-cost'])}"`
        )
      }
      return delay
    }
  },
```

**Comment by pushyamig** [comment]
@lsloan I am going to test in the morning today.

File: ccm_web/server/src/canvas/canvas.service.ts (hunk 1 above)
Comment (ssciolla) [reply ccm_web/server/src/canvas/canvas.service.ts:35]: @lsloan, before you try proactive request limiting, can you see whether this helps at all for you?

Comment (lsloan) [reply ccm_web/server/src/canvas/canvas.service.ts:35]: I'll give that a shot.

Interesting about the log messages with Canvas API header values.  I thought TS/JS would return a string like "undefined" for those.

**Comment by pushyamig** [comment]
I have tested this morning 
1. For my First attempt I tried to add 400 users to a section, 92 failed. All of them failed with 403
2.  Tried to enroll 346 users to a different section, 7 failed 
3. Tried to enroll 346 users to a different section, 21 failed

All the testing happened in about 15 minutes span

As per the code now, if user tries to attempt a bulk user enroll set (~300-400) then they might see rate limiting error, and if the user tries again with a different set of enrolling users/section (with same Token) with in next 15 min of time they might still see the 403 error since canvas is throttling the API request.

Second round of testing with 168 users enroll to 3 different sections in the same courses. In all 3 times I saw successfully enrolling users to 3 different section, the testing all in with in 5 min of time.

**Comment by ssciolla** [comment]
> I have tested this morning
> 
>     1. For my First attempt I tried to add 400 users to a section, 92 failed. All of them failed with 403
>     2. Tried to enroll 346 users to a different section, 7 failed
>     3. Tried to enroll 346 users to a different section, 21 failed

@pushyamig, did you see retries happening? I woud have expected Lance's retry settings to handle `403` cases, generally speaking.

**Comment by pushyamig** [comment]
> @pushyamig, did you see retries happening? I woud have expected Lance's retry settings to handle 403 cases, generally speaking.

Yes. I believe based on the below log messages they are happening
>beforeRetry [1] - error.response.statusCode: "403"; error.code: "undefined"
ccm_web     | 2021-10-04T15:34:04.786Z - canvas.service.ts - debug - afterResponse — "x-rate-limit-remaining": "13.677155825080035"; "x-request-cost": "undefined"
ccm_web     | 2021-10-04T15:34:04.787Z - canvas.service.ts - debug - calculateDelay [2] — "delay": "5000"; "retryOptions": "{"limit":3,"methods":["POST","GET","PUT","DELETE"],"statusCodes":[408,413,429,500,502,503,504,521,522,524,403],"errorCodes":["ETIMEDOUT","ECONNRESET","EADDRINUSE","ECONNREFUSED","EPIPE","ENOTFOUND","ENETUNREACH","EAI_AGAIN"],"maxRetryAfter":null}"; "x-rate-limit-remaining": "13.677155825080035"; "x-request-cost": "undefined";

> canvas.service.ts - debug - beforeRetry [2] - error.response.statusCode: "403"; error.code: "undefined"
ccm_web     | 2021-10-04T15:34:10.165Z - canvas.service.ts - debug - beforeRetry [2] - error.response.statusCode: "403"; error.code: "undefined"
ccm_web     | 2021-10-04T15:34:10.172Z - canvas.service.ts - debug - beforeRetry [2] - error.response.statusCode: "403"; error.code: "undefined"
ccm_web     | 2021-10-04T15:34:10.199Z - canvas.service.ts - debug - beforeRetry [2] - error.response.statusCode: "403"; error.code: "undefined"
ccm_web     | 2021-10-04T15:34:10.326Z - canvas.service.ts - debug - beforeRetry [2] - error.response.statusCode: "403"; error.code: "undefined"
ccm_web     | 2021-10-04T15:34:10.369Z - canvas.service.ts - debug - afterResponse — "x-rate-limit-remaining": "18.36509866926997"; "x-request-cost": "0.880740450969995"
ccm_web     | 2021-10-04T15:34:10.370Z - api.section.handler.ts - debug - Received response with status code 200
ccm_web     | 2021-10-04T15:34:10.425Z - canvas.service.ts - debug - afterResponse — "x-rate-limit-remaining": "33.26579473856998"; "x-request-cost": "0.843420237002249"
ccm_web     | 2021-10-04T15:34:10.425Z - api.section.handler.ts - debug - Received response with status code 200
ccm_web     | 2021-10-04T15:34:10.431Z - canvas.service.ts - debug - afterResponse — "x-rate-limit-remaining": "47.48881356835"; "x-request-cost": "0.8239842009981722"
ccm_web     | 2021-10-04T15:34:10.431Z - api.section.handler.ts - debug - Received response with status code 200
ccm_web     | 2021-10-04T15:34:10.437Z - canvas.service.ts - debug - beforeRetry [2] - error.response.statusCode: "403"; error.code: "undefined"

**Comment by ssciolla** [comment]
@pushyamig, gosh, so bizarre. I just enrolled 400 users to a bunch of different sections with minimal issues. Retries happened, but none of them had to retry a second time. I wonder if there are any `beta` versus `test` differences we have to keep in mind.

**Comment by pushyamig** [comment]
>Retries happened, but none of them had to retry a second time. I wonder if there are any beta versus test differences we have to keep in mind.

That will be bizarre.

I also see this in the Logs
` beforeRetry [3] - error.response.statusCode: "403"; error.code: "undefined"`

Along with 
` beforeRetry [2] - error.response.statusCode: "403"; error.code: "undefined"`
` beforeRetry [1] - error.response.statusCode: "403"; error.code: "undefined"`

If I understood this, the retry is happening 3 times?

File: ccm_web/server/src/canvas/canvas.service.ts (hunk 1 above)
Comment (lsloan) [reply ccm_web/server/src/canvas/canvas.service.ts]: Thanks!  That's a valuable TS lesson.

**Comment by lsloan** [comment]
# TL;DR

Checking for headers before using them helped me.  I never saw the "strange error" condition.  I increased the limit on number of retries.

# Details

As I mentioned in the commit comment for fb8c250…

>With help from @ssciolla, added some conditionals to handle infrequent error conditions:
>
>1. The response headers may not always be included in the response.  This seems to be the biggest problem that was preventing the requests from completing correctly for me.  I thought TS/JS would automatically handle these, returning the values as the string "undefined", but it does not.  Putting conditionals around the use of the headers resolved all problems for me.
>2. Sometimes the error object for `calculateDelay` wouldn't be an `HTTPError` object and wouldn't include an error code.  In those cases, Got would not compute a retry delay value.  Our logic assumed that meant it shouldn't be retried, but it turns out we got good results if we did.  This may be specific to Canvas' API.
>
>Also, in my testing of the code, found my requests hit the limit of 3 retries.  The third retry happened to be successful, but if it hadn't some of the requests would've failed.  So, I increased the limit to 4.

**Comment by lsloan** [comment]
@pushyamig, thanks for testing.  I think you were seeing results that are more like mine were last week.  I don't know why @ssciolla had an easier time than us.

Anyway, Sam found some spots in the code that could use fixing.  Mainly to not rely on assumptions.  This has made the code work for me 100% of the time.

Please test and review again.

Thank you! 🙇

**Comment by lsloan** [comment]
I don't remember changing the pipeline for this PR.  Maybe it's an automatic PR thing that I never noticed before.  Anyway, just now I moved it from backlog to in progress.

File: ccm_web/server/src/canvas/canvas.service.ts [hunk 2]
```diff
@@ -14,47 +14,62 @@ import { User } from '../user/user.model'
...
 
-interface IncomingRateLimitedCanvasHttpHeaders extends IncomingHttpHeaders {
-  'x-rate-limit-remaining'?: string
-  'x-request-cost'?: string
-}
-
 const requestorOptions: GotOptions = {
   retry: {
-    limit: 2,
+    limit: 4,
```
Comment (lsloan) [inline ccm_web/server/src/canvas/canvas.service.ts:24]: This number of retries (4) and…

File: ccm_web/server/src/canvas/canvas.service.ts [hunk 3]
```diff
@@ -14,47 +14,62 @@ import { User } from '../user/user.model'
...
     methods: ['POST', 'GET', 'PUT', 'DELETE'],
-    statusCodes: got.defaults.options.retry.statusCodes.concat([403])
+    statusCodes: got.defaults.options.retry.statusCodes.concat([403]),
+    calculateDelay: ({ attemptCount, retryOptions, error, computedValue }) => {
+      const isStrangeError = !(error instanceof HTTPError) && error.code === undefined
+      const delay = computedValue === 0 && !isStrangeError ? 0 : 5000
```
Comment (lsloan) [inline ccm_web/server/src/canvas/canvas.service.ts:29]: …this number of delay milliseconds (5000), probably should be parameterized values that are loaded from the application configuration.

I looked into doing that before my last commit.  It isn't **_very_** complex to support config parameters, but it does involve a lot of changes, especially in this file.  It was getting complicated enough that I decided it should wait until after we have a solution to the rate‑limit problem.

**Review by lsloan — COMMENTED** [review]
Could the parameterization and use of application configuration be a new issue?  I feel this PR has been open long enough.  Now that the rate‑limit problem has a solution, it would be nice to merge this PR, close the issue, and move on to another one.

File: ccm_web/server/src/canvas/canvas.service.ts (hunk 3 above)
Comment (ssciolla) [reply ccm_web/server/src/canvas/canvas.service.ts:29]: I'm fine waiting on the configuration thing; it may be nice but also seems potentially unnecessary. It would require some changes to the service itself, probably you'd want to wrap the options creation in a function and pass the necessary configuration when the `Requestor` is instantiated.

File: ccm_web/server/src/canvas/canvas.service.ts [hunk 4]
```diff
@@ -14,47 +14,62 @@ import { User } from '../user/user.model'
...
-
 const requestorOptions: GotOptions = {
   retry: {
-    limit: 2,
+    limit: 4,
     methods: ['POST', 'GET', 'PUT', 'DELETE'],
-    statusCodes: got.defaults.options.retry.statusCodes.concat([403])
+    statusCodes: got.defaults.options.retry.statusCodes.concat([403]),
+    calculateDelay: ({ attemptCount, retryOptions, error, computedValue }) => {
+      const isStrangeError = !(error instanceof HTTPError) && error.code === undefined
```
Comment (ssciolla) [inline ccm_web/server/src/canvas/canvas.service.ts:28]: Do you still think the `isStrangeError` handling is necessary? I'm not sure I ever saw it be `true` after I implemented it, maybe the incorrect `headers` accessing explains all the error cases you were seeing, I'm not sure.

Comment (lsloan) [reply ccm_web/server/src/canvas/canvas.service.ts:28]: I wasn't sure whether you saw that case being handled or not, so I left that code in place.  In my testing, I've not seen any value for that except `false`.  If you and @pushyamig don't see that case, either, then I will remove it.

**Comment by pushyamig** [comment]
I tested for the course https://umich.beta.instructure.com/courses/403334/ adding 400 users to 4 different section. I found 3 time all 400 users are added but did see failures. Below is 
Screen Shot 2021-10-05 at 10 56 21 AM
the screenshot for that

**Comment by pushyamig** [comment]
> Did you save the console output from the application when you ran the tests? If so, can you see whether the number of retries was limited? The number in the code is set to 4. Did you see any messages in the log like beforeRetry [4] or calculateDelay [4]?
Did you see any other error messages?

The order I tested for enrolling users to section is `section 20`, `section 11`, `section 10`, `section 09`, `Section 08`
These are the error message for the last run which is for adding users to Section 08.

>ccm_web     | 2021-10-05T14:55:35.505Z - canvas.service.ts - debug - calculateDelay [4] — "delay": "5000"; "isStrangeError": false; "retryOptions": "{"limit":4,"methods":["POST","GET","PUT","DELETE"],"statusCodes":[408,413,429,500,502,503,504,521,522,524,403],"errorCodes":["ETIMEDOUT","ECONNRESET","EADDRINUSE","ECONNREFUSED","EPIPE","ENOTFOUND","ENETUNREACH","EAI_AGAIN"],"maxRetryAfter":null}"; error.code: undefined"x-rate-limit-remaining": "18.49214469441995"; "x-request-cost": "undefined";
ccm_web     | 2021-10-05T14:55:35.514Z - canvas.service.ts - debug - beforeRetry [3] - error.response.statusCode: "403"; error.code: "undefined"
ccm_web     | 2021-10-05T14:55:35.520Z - canvas.service.ts - debug - beforeRetry [3] - error.response.statusCode: "403"; error.code: "undefined"

>ccm_web     | 2021-10-05T14:55:41.395Z - api.utils.ts - error - Received error status code: (403)
ccm_web     | 2021-10-05T14:55:41.396Z - api.utils.ts - error - Response body: (Response body had unexpected shape: "403 Forbidden (Rate Limit Exceeded)\n")
ccm_web     | 2021-10-05T14:55:41.396Z - api.utils.ts - error - Failed input: ({"loginId":"loganrs","type":"StudentEnrollment"})
ccm_web     | 2021-10-05T14:55:41.398Z - canvas.service.ts - debug - afterResponse — "x-rate-limit-remaining": "-4.027564406820034"; "x-request-cost": "undefined"
ccm_web     | 2021-10-05T14:55:41.398Z - canvas.service.ts - debug - calculateDelay [5] — "delay": "0"; "isStrangeError": false; "retryOptions": "{"limit":4,"methods":["POST","GET","PUT","DELETE"],"statusCodes":[408,413,429,500,502,503,504,521,522,524,403],"errorCodes":["ETIMEDOUT","ECONNRESET","EADDRINUSE","ECONNREFUSED","EPIPE","ENOTFOUND","ENETUNREACH","EAI_AGAIN"],"maxRetryAfter":null}"; error.code: undefined"x-rate-limit-remaining": "-4.027564406820034"; "x-request-cost": "undefined";
ccm_web     | 2021-10-05T14:55:41.399Z - api.utils.ts - error - Received error status code: (403)
ccm_web     | 2021-10-05T14:55:41.399Z - api.utils.ts - error - Response body: (Response body had unexpected shape: "403 Forbidden (Rate Limit Exceeded)\n")

>ccm_web     | 2021-10-05T14:55:35.980Z - canvas.service.ts - debug - calculateDelay [4] — "delay": "5000"; "isStrangeError": false; "retryOptions": "{"limit":4,"methods":["POST","GET","PUT","DELETE"],"statusCodes":[408,413,429,500,502,503,504,521,522,524,403],"errorCodes":["ETIMEDOUT","ECONNRESET","EADDRINUSE","ECONNREFUSED","EPIPE","ENOTFOUND","ENETUNREACH","EAI_AGAIN"],"maxRetryAfter":null}"; error.code: undefined"x-rate-limit-remaining": "8.982203255460036"; "x-request-cost": "undefined";
ccm_web     | 2021-10-05T14:55:35.985Z - canvas.service.ts - debug - afterResponse — "x-rate-limit-remaining": "-33.639951020520016"; "x-request-cost": "0.2984218630143687"
ccm_web     | 2021-10-05T14:55:35.986Z - api.section.handler.ts - debug - Received response with status code 200
ccm_web     | 2021-10-05T14:55:35.986Z - canvas.service.ts - debug - afterResponse — "x-rate-limit-remaining": "24.21252445483003"; "x-request-cost": "0.29641698498926417"
ccm_web     | 2021-10-05T14:55:35.986Z - api.section.handler.ts - debug - Received response with status code 200
ccm_web     | 2021-10-05T14:55:35.986Z - canvas.service.ts - debug - afterResponse — "x-rate-limit-remaining": "23.832966111899964"; "x-request-cost": "0.32675153501600107"
ccm_web     | 2021-10-05T14:55:35.986Z - api.section.handler.ts - debug - Received response with status code 200
ccm_web     | 2021-10-05T14:55:35.988Z - canvas.service.ts - debug - afterResponse — "x-rate-limit-remaining": "9.142898078470012"; "x-request-cost": "undefined"
ccm_web     | 2021-10-05T14:55:35.989Z - canvas.service.ts - debug - calculateDelay [4] — "delay": "5000"; "isStrangeError": false; "retryOptions": "{"limit":4,"methods":["POST","GET","PUT","DELETE"],"statusCodes":[408,413,429,500,502,503,504,521,522,524,403],"errorCodes":["ETIMEDOUT","ECONNRESET","EADDRINUSE","ECONNREFUSED","EPIPE","ENOTFOUND","ENETUNREACH","EAI_AGAIN"],"maxRetryAfter":null}"; error.code: undefined"x-rate-limit-remaining": "9.142898078470012"; "x-request-cost": "undefined";

We can chat over slack by DM if this is not clear to you

**Comment by lsloan** [comment]
@pushyamig, thanks, that's helpful and enlightening.

I see that you had the default limit of 4, but in at least one of the log messages, it shows that it was retry number 5.  I didn't know that was how the Got logic works.  That tells us that at least one request was tried up to the limit (4 times), but even that last retry failed.  It would've tried again if the limit hadn't been reached.

I'd like to know what would happen for you if the limit were increased to a large number, like 10.

As @ssciolla mentioned earlier, there may be differences between the Canvas "test" and "beta" instances.  I expect that "test" would be similar to production, which is why I use that for development and testing.  I think "beta" has features that are not in production or may never reach production.  We might not know how those features would affect our development or whether it would ever affect production if the features are ever released.

So, I'll try using the "beta" environment myself and see what's happening.

**Comment by pushyamig** [comment]
> I expect that "test" would be similar to production, which is why I use that for development and testing. I think "beta" has features that are not in production or may never reach production. We might not know how those features would affect our development or whether it would ever affect production if the features are ever released.

I know certain time difference b/w test/best and Prod is processing power. Like Canvas Job Server is very slow some times, loading time of the app is slow. I don't know if they have difference with API (rate limiting, response time).  

Beta might have latest features preview for Functional analysts of canvas. To me the closest match of prod is the dev instance( which we won't make use of it)

@melindakraft and @jonespm   Thoughts on difference between prod/test/beta/dev.

**Comment by lsloan** [comment]
My point is that it's probably not safe to group the two environments together as "test/beta".

It makes sense to group "test/prod", because the features of the "test" and "prod" environments are supposed to be the same by definition.  The differences between those two should only be content and processing power.  There should **_not_** be any difference in features, because "test" is the environment where we get to try things out (i.e., test them) before using them in "prod".

The "beta" environment may have new features not present in "test/prod".  That's the common definition of a "beta" environment.  Those features may affect the code we're developing and testing.  That could be helpful if the new features that affect our code make it to "prod" someday, but if the new features are dropped/changed, then we may have been taking unnecessary time to optimize our code for conditions that will never occur.

I think that we should standardize (or at least, temporarily agree) on which environment we are using.  If @ssciolla and I are using "test" for development because it is most similar to "prod", then the whole team should be using that, at least for this project.  If we get feedback that it doesn't work on "beta", it might not even be relevant to us.

File: ccm_web/server/src/canvas/canvas.service.ts (hunk 4 above)
Comment (ssciolla) [reply ccm_web/server/src/canvas/canvas.service.ts:28]: I didn't see it in my testing again either. I'm gonna say remove it. Maybe I was getting confused, and the issue was wholly a `headers` one.

**Review by ssciolla — APPROVED** [review]
I did another round of testing with your code, and did not see any issues. I approve this, dependent on you removing the `isStrangeError` stuff. We can still probably diagnose situations like this with the logging you are doing now.

Note on merging: *Please* use "squash" (or "rebase" if you must have separate commits), so that the commit history is easier to read. (I noticed things got messy after the enrollment PR was merged.)

**Comment by pushyamig** [comment]
> If @ssciolla and I are using "test" for development because it is most similar to "prod", then the whole team should be using that, at least for this project. 

Test and Prod are not the same since the canvas URL is different one has vanity and another ends with instructure.com. based the environment you launch from there are certain things different. For example I saw this CSP error in [beta](https://github.com/tl-its-umich-edu/canvas-course-manager-next/issues/103) not in test. Since prod is not a vanity URL so I was making sure I test from beta so to look for issues that are related to URL. SameSite setting won't have any effect if ending with `umich.edu` and I was making sure to test on beta to make sure we catch error of those sorts.

I really don't have preference here beta/test. I had a different purpose for testing in beta. It's is really a surprise to me canvas throttling is different based on canvas instance(test/beta). I only test in Test for throttling if team want me to do so.

**Comment by ssciolla** [comment]
I'm not sure how to handle the environment differences here. One thing we could try (which I suggested to @lsloan) is batching requests (using `Promise.all` on some number of promises), and then starting a new batch once the previous batch is all resolved. Maybe this would minimize the load placed on the API at the start, and slow down the clock so to speak on the rate limiting.

File: ccm_web/server/src/canvas/canvas.service.ts [hunk 5]
```diff
@@ -14,47 +14,62 @@ import { User } from '../user/user.model'
...
+      const delay = computedValue === 0 && !isStrangeError ? 0 : 5000
+
+      let logMessage: string =
+        `calculateDelay [${String(attemptCount)}] — ` +
+        `"delay": "${String(delay)}"; ` +
+        `"isStrangeError": ${String(isStrangeError)}; ` +
+        `"retryOptions": "${JSON.stringify(retryOptions)}"; ` +
+        `error.code: ${String(error.code)}`
```
[the remark this answers is not in the document]
Comment (lsloan) [reply ccm_web/server/src/canvas/canvas.service.ts:36]: True!  The `error.code` line was added by you, but I should've corrected the quotes before committing it.  I _thought_ I had done that.  Maybe I had changed only one of the other log messages.

Some the log messages in the hooks need this correction, too.

**Comment by lsloan** [comment]
@ssciolla, thank you very much for the code review and approval.  I'll make some changes and remove the `isStrangeError`-related code shortly.

# Environment differences

WRT "test" vs. "beta" environments for development, we should use whichever one is most similar to "prod".  I acknowledge there are URL differences among the three environments.  Those may contribute to security handling issues.  However, once those are overcome, what other differences are there between the three environments?

My contention is that aside from URL differences, "test" and "prod" functionally should be very similar, if not identical.  The "beta" environment is likely to have features not found in the other environments.  We may or may not know what those features are.

# Testing in "beta", summary

Disregarding the differences between the three environments, I decided to test the code in the "beta" environment myself. I created six sections and enrolled 400 users in each of them.  They were **_100%_** successful. 🎉 

# Testing in "beta", more detail

The first section I tested with completed without any retries.  I remembered that @pushyamig was successful in that environment sometimes, too, so I continued testing.

I created five more sections in a row, all with sequential ID numbers, so I could test them quickly one after another.  Sam and I had done that before in the "test" environment.  We noticed that when doing one section at a time, we were more likely to see success, but if we processed multiple sections in as rapid succession as we could, that would result in more failed requests and subsequent retries.

As expected, there were more failures and retries with this testing method.  Of the five sections I tested with in this manner, three of them contained retries.  No queries were retried more than **_once_**.  Of all the three sections that included retries, the total number of retries was only **_ten_**…

> beta Log at 2021-10-05 2-32-13 PM.txt:ccm_web     | 2021-10-05T18:31:18.908Z - canvas.service.ts - debug - beforeRetry [1] - error.response.statusCode: "undefined"; error.code: "ECONNRESET"
beta Log at 2021-10-05 2-34-21 PM.txt:ccm_web     | 2021-10-05T18:33:59.056Z - canvas.service.ts - debug - beforeRetry [1] - error.response.statusCode: "403"; error.code: "undefined"
beta Log at 2021-10-05 2-34-21 PM.txt:ccm_web     | 2021-10-05T18:33:59.166Z - canvas.service.ts - debug - beforeRetry [1] - error.response.statusCode: "403"; error.code: "undefined"
beta Log at 2021-10-05 2-34-21 PM.txt:ccm_web     | 2021-10-05T18:33:59.172Z - canvas.service.ts - debug - beforeRetry [1] - error.response.statusCode: "403"; error.code: "undefined"
beta Log at 2021-10-05 2-34-21 PM.txt:ccm_web     | 2021-10-05T18:33:59.178Z - canvas.service.ts - debug - beforeRetry [1] - error.response.statusCode: "403"; error.code: "undefined"
beta Log at 2021-10-05 2-34-21 PM.txt:ccm_web     | 2021-10-05T18:33:59.182Z - canvas.service.ts - debug - beforeRetry [1] - error.response.statusCode: "403"; error.code: "undefined"
beta Log at 2021-10-05 2-34-21 PM.txt:ccm_web     | 2021-10-05T18:33:59.234Z - canvas.service.ts - debug - beforeRetry [1] - error.response.statusCode: "403"; error.code: "undefined"
beta Log at 2021-10-05 2-34-21 PM.txt:ccm_web     | 2021-10-05T18:33:59.236Z - canvas.service.ts - debug - beforeRetry [1] - error.response.statusCode: "403"; error.code: "undefined"
beta Log at 2021-10-05 2-34-21 PM.txt:ccm_web     | 2021-10-05T18:33:59.240Z - canvas.service.ts - debug - beforeRetry [1] - error.response.statusCode: "403"; error.code: "undefined"
beta Log at 2021-10-05 2-35-13 PM.txt:ccm_web     | 2021-10-05T18:34:50.965Z - canvas.service.ts - debug - beforeRetry [1] - error.response.statusCode: "undefined"; error.code: "ECONNRESET"

# Conclusion

My tests in "beta" were successful.  Others' results may vary.  Last week when Sam and I were both testing in "test", I was getting errors that he did not.  As we know now, it may have been due to poor logging and inappropriate handling of missing response headers.  However, when he continued working on the problem later on Friday evening, he began to see problems as well.  So one tester may see problems that others do not at any given time.

**Comment by lsloan** [comment]
@pushyamig: I've committed the changes I wanted to make.  It's ready for you to test in the Canvas-test environment.

**Comment by pushyamig** [comment]
I tested in the Canvas test enrolling user to 5 different section in the same course https://canvas-test.it.umich.edu/courses/403334/. Same user set. I saw retries happening from 1-4 times.  All are success this time. 

Screen Shot 2021-10-05 at 4 18 44 PM

**Review by pushyamig — APPROVED** [review]
Looks Like we have a solution for enrolling 400 user set and we are going to keep that limit. Hopefully things are faster in Prod. I saw the entire process to about 20 sec which is a good from user perspective. 

Nice work @lsloan ! We can use is approach for Add Non-UMich user and decide a limit for that use case

**Comment by lsloan** [comment]
Thanks for the review and testing, @pushyamig.  I'm glad you were able to get good results.

I'll merge this PR shortly.
